docs/adrs/0262-custom-label-support.md

```diff
@@ -0,0 +1,43 @@
+# ADR 262: Custom Label Support
+**Date**: 2020-01-14
+
+**Status**: Proposed
+
+## Context
+
+In the case of a single repo having multiple self-hosted runners, actions do not currently support any way of selecting which runner or kind of runner a job should run on, other than by architecture or OS.
+
+Example use cases for this could be:
+
+* Your project involves some machine learning models. You want to add a second runner with a GPU for testing your models, and to have your model tests job use this runner, but other jobs to use the runner without a GPU. 
+* You want to use GitHub actions to deploy your app, and for security you use a self-hosted runner inside your isolated network. If you have a separate isolated network for each environment (e.g. testing, staging, production), you would need to run your deploy job on a different runner depending on which environment you are deploying to.
+* Any other situation where you would like to select which self-hosted runner to use for a job based on the capabilities or characteristics of the runner.
+
+GitHub actions already support selecting a job runner based on labels, and automatically give runners labels based on OS and arch, so a natural extension would be to allow operators of self-hosted runners to add their own labels, so that users can select job runners based on any operator-defined characteristic.
+
+See Issue: https://github.com/actions/runner/issues/262
+
+## Decision
+
+This ADR proposes that we add a `--label` option to `config.sh`, which could be used to add custom additional labels to the configured runner.
```

Comment:
Should we add a file option as well like we have for ENV?

Reply:
Yeah, I like that as well.  If we do it should be by convention like env.  That means create a file name labels (or is it .labels?) - whatever env is - and it just works.  We should apply envvars first, and then process each line in the file for envvars (${HOSTNAME}) and trim.

Reply:
Should we consider only have one way of doing this? I prefer file over cmdline args. 😄

Reply:
I really like the file idea. Can you point me to an example of this working for envvars?

Reply:
From my perspective in docker/kubernetes-land, it's easier to automate config using commandline options than using files. When deploying a container, changing the commandline args is just a change in the deployment config, but changing a file involves either rebuilding your container image and baking-in the settings or writing a startup wrapper script that generates the file from env vars/commandline.
Not so sure about hardcoding the path to the file, but I like having the option to load from a file. Docker similarly has a `docker run --label-file path/to/labels.txt ...` option which loads a newline-seperated set of labels. Using this, an organisation could make a VM/container image with their different labels files in it (say `labels.testing.txt`, `labels.staging.txt`, `labels.prod.txt`) and select which one to use with `--label-file` at deploy time.

Reply:
@bryanmacfarlane 
If you want to support other ways to set the settings then I'd model it after something like https://github.com/spf13/viper - not that you'd use that library but you might want to look at how it loads settings. 

A command line argument is kinda a necessity IMO. It would be bizarre not to have that option.

Reply:
A bit of background on why we have a file for envvars.  When the runner is run as a service the service file is generated (linux or mac) which runs svc.sh and we didn't want users to have to edit the file (post config).  There's also scenarios to modify and just restart the svc.  All that said, it's post config modifications to the env which really doesn't apply here.  So, if we couple that with Joe's principle of just having one, then that means we can just have cmd line config option and that's it.  There's no strong scenario like there is for runtime envvars and restart since config is a one time only operation.  The container point is also a good one where it can just be passed through (instead of having to modify files).  So I think taking all that into consideration means only cmd line is good.

Reply:
We should be consistent within the runner and offer cmd line args and picking up from a file similar to our `.env`.  Please add to the ADR.

Reply:
I'm also not proposing or referring to other ways to potentially set settings.  I'm outlining how the existing runner applies settings from file, envvars, and then override with args or prompt.  e.g. we're not rethinking the settings system here - we're just adding one more arg.